File: src/main/java/pl/edu/pw/elka/mtoporow/cevolver/lib/model/matrix/TMatrix.scala
package pl.edu.pw.elka.mtoporow.cevolver.lib.model.matrix

import org.apache.commons.math3.complex.Complex
import org.apache.commons.math3.linear.{Array2DRowFieldMatrix, FieldMatrix}

/**
 * Macierz T
 * Data utworzenia: 14.06.15, 19:01
 * @author Michał Toporowski
 */
class TMatrix(val t11: Complex, val t12: Complex, val t21: Complex, val t22: Complex) {

  def asMatrix: FieldMatrix[Complex] = new Array2DRowFieldMatrix[Complex](Array(Array(t11, t12), Array(t21, t22)))

  def toSMatrix: SMatrix = {
    val s11 = t12.divide(t22)
    val s12 = t11.multiply(t22).subtract(t12.multiply(t21)).divide(t22)
    val s21 = Complex.ONE.divide(t22)
    val s22 = t21.negate().divide(t22)
    new SMatrix(s11, s12, s21, s22)
  }

  /**
   * Mnoży macierze T (połączenie kaskadowe)
   *
   * @param that
   * @return
   */
  def *(that: TMatrix) = {
    val matrix = asMatrix.multiply(that.asMatrix).getData
    new TMatrix(matrix(0)(0), matrix(0)(1), matrix(1)(0), matrix(1)(1))
  }

  /**
   * Zwraca element s11 bez obliczania pozostałych elementów macierzy S
   */
  def s11: Complex = {
    t12.divide(t22)
  }
}

File: src/main/java/pl/edu/pw/elka/mtoporow/cevolver/lib/model/matrix/TMatrix.java
package pl.edu.pw.elka.mtoporow.cevolver.lib.model.matrix;

import org.apache.commons.math3.complex.Complex;
import org.apache.commons.math3.linear.Array2DRowFieldMatrix;
import org.apache.commons.math3.linear.FieldMatrix;

/**
 * Implementacja macierzy T
 * Data utworzenia: 29.01.16, 16:38
 *
 * @author Michał Toporowski
 */
public class TMatrix {
    private final FieldMatrix<Complex> matrix;

    public TMatrix(Complex t11, Complex t12, Complex t21, Complex t22) {
        this(new Array2DRowFieldMatrix<>(new Complex[][]{{t11, t12}, {t21, t22}}));
    }

    private TMatrix(FieldMatrix<Complex> matrix) {
        this.matrix = matrix;
    }

    /**
     * Mnoży tę macierz przez inną
     *
     * @param other inna macierz T
     * @return macierz wynikowa tj. macierz reprezentująca połączenie kaskadowe czwórników
     */
    public TMatrix multiply(TMatrix other) {
        return new TMatrix(this.matrix.multiply(other.matrix));
    }

    /**
     * Zwraca element S11 macierzy reprezentującej połączenie kaskadowe tej macierzy i macierzy reprezentowanej
     * jedynie przez element s11'
     *
     * @param s11 s11'
     * @return element s11 wynikowej macierzy
     */
    public Complex getS11WithCascadeS11(final Complex s11) {
        // TODO:: czy to dobrze?
        // s11m = (t11 * s11' + t12) / (t21 + s11' + t22)
        return matrix.getEntry(0, 0).multiply(s11).add(matrix.getEntry(0, 1))
                .divide(matrix.getEntry(1, 0).multiply(s11).add(matrix.getEntry(1, 1)));
    }

    /**
     * Zwraca element S11 z macierzy S przekształconej z tej macierzy T
     *
     * @return S11 = T12 / T22 jako liczba zespolona
     */
    public Complex getS11() {
        return matrix.getEntry(0, 1).divide(matrix.getEntry(1, 1));
    }

    /**
     * Zwraca macierz jednostkową
     *
     * @return jednostkowa macierz T
     */
    public static TMatrix identity() {
        return new TMatrix(Complex.ONE, Complex.ZERO, Complex.ZERO, Complex.ONE);
    }
}
